### PR TITLE
Make exec package agnostic of platform

### LINKED ISSUE #6160: Make pkg/sync platform agnostic
/kind feature

The pkg/sync package is for the moment designed for working on Kubernetes platform only.

## Which functionality do you think we should add?

Make the package platform agnostic.

## Why is this needed?

So we can reuse it for other platforms (Podman)

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind code-refactoring

**What does this PR do / why we need it:**

This package makes the exec package, and thus the sync package using it, agnostic of the platform. 
 The platform is injected by the dependency injection system, depending on the `--run-on` flag.

**Which issue(s) this PR fixes:**

Fixes #6160 

**PR acceptance criteria:**

- [ ] Unit test 

- [ ] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

File: pkg/sync/copy.go [hunk 1]
```diff
@@ -43,14 +45,32 @@ func CopyFile(extracter SyncExtracter, localPath string, compInfo ComponentInfo,
...
+	klog.V(3).Infof("Executing command %s", strings.Join(cmdArr, " "))
+	err := a.platformClient.ExecCMDInContainer(containerName, podName, cmdArr, &stdout, &stderr, stdin, false)
+	if err != nil {
+		log.Errorf("Command '%s' in container failed.\n", strings.Join(cmdArr, " "))
+		log.Errorf("stdout: %s\n", stdout.String())
+		log.Errorf("stderr: %s\n", stderr.String())
+		log.Errorf("err: %s\n", err.Error())
+		return err
+	}
+	return nil
```
Comment (valaparthvi) [inline pkg/sync/copy.go:71]:
```suggestion
	}
	return err
```

File: pkg/sync/sync.go [hunk 2]
```diff
@@ -10,25 +10,25 @@ import (
...
 
 	"github.com/redhat-developer/odo/pkg/exec"
-	"github.com/redhat-developer/odo/pkg/kclient"
+	"github.com/redhat-developer/odo/pkg/platform"
 	"github.com/redhat-developer/odo/pkg/util"
 
 	"k8s.io/klog"
 )
 
 // SyncClient is a Kubernetes implementationn for sync
```
Comment (valaparthvi) [inline pkg/sync/sync.go:19]:
```suggestion
// SyncClient is a Kubernetes implementation for sync
```

Or maybe something more appropriate.

**Review by valaparthvi — CHANGES_REQUESTED** [review]

**Review by valaparthvi — APPROVED** [review]

**Review by rm3l — COMMENTED** [review]
/approve

**Review by valaparthvi — APPROVED** [review]

**Comment by valaparthvi** [comment]
/override Kubernetes-Integration-Tests/Kubernetes-Integration-Tests
/override OpenShift-Integration-tests/OpenShift-Integration-tests
/override ci/prow/v4.11-integration-e2e
Overriding the tests. All point to the same failure.
```
Summarizing 2 Failures:
  [FAIL] E2E Test starting with non-empty Directory [It] should verify developer workflow from non-empty Directory
  /go/src/github.com/redhat-developer/odo/tests/helper/helper_interactive.go:111
  [FAIL] E2E Test starting with empty Directory [It] should verify developer workflow from empty Directory
  /go/src/github.com/redhat-developer/odo/tests/helper/helper_generic.go:76

Ran 6 of 6 Specs in 162.545 seconds
FAIL! -- 4 Passed | 2 Failed | 0 Pending | 0 Skipped
```

@feloy I think it's alright to merge this PR given the errors. But I am going to put this on hold, in case you'd like to wait until #6243 goes in. Feel free to remove the hold if you want it to go in right away.

/hold

**Comment by feloy** [comment]
/override NoCluster-Integration-tests/NoCluster-Integration-tests
Not applicable

**Pull request merged**